### PR TITLE
Add objects attributes introspection to display them

### PR DESCRIPTION
**Pull request by Delgan**
@Qix- Here is my solution for implementing #71. 🙂 

The code is not as straightforward as I wished, but it should be comprehensive enough. If you have alternative ideas or others proposals for the implementation, don't hesitate to let me know. I tried to think as many edge cases as possible. I also used [`inspect.getattr_static()`](https://docs.python.org/3/library/inspect.html#inspect.getattr_static) as you suggested. 👍

### CODE REVIEW

File: test/output/python-dumb-UTF-8-nocolor.out [hunk 1]
```diff
@@ -306,3 +312,25 @@ TypeError: unsupported operand type(s) for +: 'int' and 'bytes'
...
+    └ <function foo at 0xDEADBEEF>
+  File "test/test_attributes.py", line 15, in foo
+    ... + 1 + bar(a).b + a.forbidden + a.nope.a + x.__bool__ or a. b . isdigit() and .3 + ...
+                  │      │ │           │          │ │           │  │   └ <method 'isdigit' of 'str' objects>
+                  │      │ │           │          │ │           │  └ '123'
+                  │      │ │           │          │ │           └ <__main__.Obj object at 0xDEADBEEF>
+                  │      │ │           │          │ └ <slot wrapper '__bool__' of 'NoneType' objects>
+                  │      │ │           │          └ None
+                  │      │ │           └ <__main__.Obj object at 0xDEADBEEF>
+                  │      │ └ <property object at 0xDEADBEEF>
```
Comment (alexmojaki) [inline test/output/python-dumb-UTF-8-nocolor.out]: I'm writing some code related to this and I just realised this is a problem. Users don't care about descriptors and getattr_static. If the traceback says `a.forbidden` is a `property object` then they'll think that's what it literally evaluates to when it doesn't. What we need is not just the result of `getattr_static` but rather the actual result of `getattr` if we know it can be evaluated safely or nothing at all. I can't see any easy way to do this. The code for `getattr_static` is quite complex and accounts for lots of edge cases, showing that this problem space is tricky. My best idea is to keep using `getattr_static` but assume it's wrong (and thus not show it) if it has `__get__` or `__set__` attributes. Even that is probably not as easy as it sounds to do correctly.

Comment (Delgan) [reply test/output/python-dumb-UTF-8-nocolor.out]: > Users don't care about descriptors and getattr_static

I agree.

> they'll think that's what it literally evaluates to when it doesn't

I think they will mostly don't care, but I agree it's unfortunate that there is no way to differentiate between the two.

```python
import inspect

def foo(self):
    return 1

foop = property(foo)

class Bar:
    a = foop

bar = Bar()
bar.b = foop

print(bar.a)  # 1
print(bar.b)  # < property object at 0x765fe090>

print(inspect.getattr_static(bar, "a"))  # < property object at 0x765fe090>
print(inspect.getattr_static(bar, "b"))  # < property object at 0x765fe090>
```

> What we need is not just the result of getattr_static but rather the actual result of getattr if we know it can be evaluated safely or nothing at all.

There is as far as I know absolutely no way to know if this can be evaluated "safely" or not. It is required to run the decorated method to retrieve the result, but as it is likely to generate unexpected side effects, it thus can't be considered safe. This is why it was discussed to only try to fetch the "static" attributes.

> My best idea is to keep using the result of getattr_static but assume it's wrong (and thus not show it) if it has __get__ or __set__ attributes. Even that is probably not as easy as it sounds to do correctly.

It could be interesting if you manage to find a safe heuristic.
However, I'm not sure it's worth it, it seems complicated for little benefit. It seems such introspection can't be made perfect, using `getattr_static()` seems to be a good approximation to make it "good enough" by solving the main use cases.
Also, by deciding to not display some attributes while others are, it may lead users to wonder why they are not seeing the values of some properties, while printing `<property object at 0xDEADBEEF>` is at least explicit.

**Comment by Qix-** [comment]
@delgan make sure to sign into issuehunt and claim this so I can award the bounty :)

**Comment by Delgan** [comment]
> @Delgan make sure to sign into issuehunt and claim this so I can award the bounty :)

@Qix- Thanks for supporting OSS contributions like this! However, I'm not interested by the bounty, I would prefer you keep it for all the work done so far, or this to be reinvested in another issue.

File: test/output/python-dumb-UTF-8-nocolor.out (hunk 1 above)
Comment (alexmojaki) [reply test/output/python-dumb-UTF-8-nocolor.out:330]: > There is as far as I know absolutely no way to know if this can be evaluated "safely" or not. It is required to run the decorated method to retrieve the result, but as it is likely to generate unexpected side effects, it thus can't be considered safe. This is why it was discussed to only try to fetch the "static" attributes.

Right, what I meant was that if you have to run some arbitrary code such as a descriptor method, assume it's unsafe and don't run it.

In the end I took the source of getattr_static and modified it to implement what I described above - either get the attribute if it's definitely safe, or not at all. This is in my library https://github.com/alexmojaki/pure_eval which in turn is used in my own traceback library https://github.com/alexmojaki/stack_data. This is old news now but I suddenly remembered this and thought you might be interested.

Comment (Delgan) [reply test/output/python-dumb-UTF-8-nocolor.out:330]: Haha, that's funny to see this discussion resurface. :smile: 
I looked at the implementation and it seems to use very advanced techniques. Thanks for sharing, it's always interesting to see how to push the internal limits of Python introspection. :+1:

Comment (Qix-) [reply test/output/python-dumb-UTF-8-nocolor.out:330]: @alexmojaki Man that's a crazy approach to this, well done. I always love seeing things like this, thanks for thinking of us years later! 😄
